cpp/tensorrt_llm/kernels/communicationKernels/allReduceFusionKernels.h

```diff
@@ -0,0 +1,137 @@
+/*
+ * Copyright (c) 2022-2024, NVIDIA CORPORATION.  All rights reserved.
+ *
+ * Licensed under the Apache License, Version 2.0 (the "License");
+ * you may not use this file except in compliance with the License.
+ * You may obtain a copy of the License at
+ *
+ *     http://www.apache.org/licenses/LICENSE-2.0
+ *
+ * Unless required by applicable law or agreed to in writing, software
+ * distributed under the License is distributed on an "AS IS" BASIS,
+ * WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+ * See the License for the specific language governing permissions and
+ * limitations under the License.
+ */
+
+#pragma once
+#include <NvInferRuntime.h>
+#include <cuda_bf16.h>
+#include <cuda_fp16.h>
+
+#include "tensorrt_llm/common/assert.h"
+#include "tensorrt_llm/common/cudaUtils.h"
+#include "tensorrt_llm/runtime/ipcUtils.h"
+
+namespace tensorrt_llm::kernels::ar_fusion
+{
+template <typename DType>
+struct ElemsPerAccess;
+
+template <>
+struct ElemsPerAccess<half>
+{
+    static constexpr int value = 8;
+};
+
+template <>
+struct ElemsPerAccess<nv_bfloat16>
+{
+    static constexpr int value = 8;
+};
+
+template <>
+struct ElemsPerAccess<float>
+{
+    static constexpr int value = 4;
+};
+
+template <typename DType>
+static constexpr int kElemsPerAccess = ElemsPerAccess<DType>::value;
+static constexpr int kOneShotMaxToken = 128;
+static constexpr int kBarrierFlagCount = 256;
+
+enum class AllReduceFusionPattern : int
+{
+    kAllReduce = 0,
+    kARResidualRMSNorm = 1,
+    kARResidualRMSNormFP8Quant = 2,
+    kARResidualRMSNormFP4Quant = 3,
+    // The difference between the V2 and the standard version is that the V2 version outputs the result of the norm.
+    kARResidualRMSNormFP8QuantV2 = 4,
+    kARResidualRMSNormFP4QuantV2 = 5
```

Comment:
```suggestion
    // The difference between these two and the standard version is that the NormOut version outputs the result of the norm.
    kARResidualRMSNormOutFP8Quant = 4,
    kARResidualRMSNormOutFP4Quant = 5
```
V2 is too confusing.

Reply:
fixed